### PR TITLE
`rust-server-codegen`: add support for `@httpPayload` in requests

### PR DESCRIPTION
**Pull request by david-perez**
This commit adds support for the `httpPayload` Smithy trait when applied
to operation input members (request deserialization).

The code to deserialize HTTP-bound data from HTTP responses in
`ResponseBindingGenerator.kt` has been moved into a common class,
`HttpBindingGenerator.kt`, since it's useful for both clients and
servers in deserializing data from HTTP requests and responses,
respectively.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpProtocolGenerator.kt [hunk 1]
```diff
@@ -548,18 +552,56 @@ private class ServerHttpProtocolImplGenerator(
...
+            HttpLocation.PAYLOAD -> {
+                val docShapeHandler: RustWriter.(String) -> Unit = { body ->
+                    rust(
+                        "#T($body)",
+                        structuredDataParser.documentParser(operationShape),
+                    )
+                }
+                val structureShapeHandler: RustWriter.(String) -> Unit = { body ->
+                    rust("#T($body)", structuredDataParser.payloadParser(binding.member))
+                }
```
Comment (crisidev) [inline codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpProtocolGenerator.kt:574]: Out of curiosity, do you really need to explicitly set the types here? We don't really do it anywhere..

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpProtocolGenerator.kt [hunk 2]
```diff
@@ -548,18 +552,56 @@ private class ServerHttpProtocolImplGenerator(
...
+                val deserializer = httpBindingGenerator.generateDeserializePayloadFn(
+                    operationShape,
+                    binding,
+                    errorSymbol,
+                    docHandler = docShapeHandler,
+                    structuredHandler = structureShapeHandler
+                )
+                return if (binding.member.isStreaming(model)) {
+                    writable { rust("""unimplemented!("streaming request bodies");""") }
```
Comment (crisidev) [inline codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpProtocolGenerator.kt:583]: I think this should be a `todo!()` to be consistent with the rest..

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/generators/http/HttpBindingGenerator.kt [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+ *     - deserializing data from an HTTP response (we are a client).
+ */
+enum class HttpMessageType {
+    REQUEST, RESPONSE
+}
+
+/**
+ * This class generates Rust functions that deserialize data from an HTTP message.
+ * They are useful for *both*:
+ *     - servers (that deserialize HTTP requests); and
```
Comment (crisidev) [inline codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/generators/http/HttpBindingGenerator.kt:62]:
```suggestion
 *     - servers (that deserialize HTTP requests).
```

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/generators/http/HttpBindingGenerator.kt [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
+import software.amazon.smithy.rust.codegen.util.dq
+import software.amazon.smithy.rust.codegen.util.hasTrait
+import software.amazon.smithy.rust.codegen.util.isPrimitive
+import software.amazon.smithy.rust.codegen.util.isStreaming
+import software.amazon.smithy.rust.codegen.util.toSnakeCase
+
+/**
+ * The type of HTTP message from which we are deserializing the HTTP-bound data.
+ * We are either:
+ *     - deserializing data from an HTTP request (we are a server); or
```
Comment (crisidev) [inline codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/generators/http/HttpBindingGenerator.kt:52]:
```suggestion
 *     - deserializing data from an HTTP request (we are a server).
```

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/generators/http/HttpBindingGenerator.kt [hunk 5]
```diff
@@ -0,0 +1,383 @@
...
+ * This class generates Rust functions that deserialize data from an HTTP message.
+ * They are useful for *both*:
+ *     - servers (that deserialize HTTP requests); and
+ *     - clients (that deserialize HTTP responses)
+ * because the deserialization logic is entirely similar. In the cases where they _slightly_ differ,
+ * the [HttpMessageType] enum is used to distinguish.
+ *
+ * For deserialization logic that is wholly different among clients and servers, use the classes:
+ *     - [ServerRequestBindingGenerator] from the `codegen-server` project for servers; and
+ *     - [ResponseBindingGenerator] from this project for clients
```
Comment (crisidev) [inline codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/generators/http/HttpBindingGenerator.kt:69]:
```suggestion
 * For deserialization logic that is wholly different among clients and servers, use the classes:
 *     - [ServerRequestBindingGenerator] from the `codegen-server` project for servers.
 *     - [ResponseBindingGenerator] from this project for clients.
```

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/generators/http/HttpBindingGenerator.kt [hunk 6]
```diff
@@ -0,0 +1,383 @@
...
+    private val index = HttpBindingIndex.of(model)
+    private val headerUtil = CargoDependency.SmithyHttp(runtimeConfig).asType().member("header")
+    private val defaultTimestampFormat = TimestampFormatTrait.Format.EPOCH_SECONDS
+    private val dateTime = RuntimeType.DateTime(runtimeConfig).toSymbol().rustType()
+    private val httpSerdeModule = RustModule.private("http_serde")
+
+    /**
+     * Generate a function to deserialize [binding] from HTTP headers
+     *
+     * The name of the resulting function is returned as a String
```
Comment (crisidev) [inline codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/generators/http/HttpBindingGenerator.kt:91]:
```suggestion
     * Generates a function to deserialize [binding] from HTTP headers.
     *
     * The name of the resulting function is returned as a String.
```

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/generators/http/HttpBindingGenerator.kt [hunk 7]
```diff
@@ -0,0 +1,383 @@
...
+        docShapeHandler: RustWriter.(String) -> Unit,
+        httpMessageType: HttpMessageType = HttpMessageType.RESPONSE
+    ) {
+        val member = binding.member
+        val targetShape = model.expectShape(member.target)
+        // There is an unfortunate bit of dual behavior caused by an empty body causing the output to be `None` instead
+        // of an empty instance of the response type.
+        withBlock("(!body.is_empty()).then(||{", "}).transpose()") {
+            when (targetShape) {
+                // TODO What if targetShape is set/map/list ?
```
Comment (crisidev) [inline codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/generators/http/HttpBindingGenerator.kt:251]: This fails the CI: https://github.com/awslabs/smithy-rs/runs/4778596089?check_suite_focus=true#step:4:73

**Review by crisidev — APPROVED** [review]
Minor comments. LGTM.

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/generators/http/HttpBindingGenerator.kt (hunk 7 above)
Comment (david-perez) [reply codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/generators/http/HttpBindingGenerator.kt:251]: Ah, was reading the code and forgot to test that.

Turns out that if you try to target a set/map/list, you get:

```
ERROR: com.amazonaws.simple#HealthcheckInputRequest$payload (ProtocolHttpPayload)
     @ /local/home/davidpz/workplace/smithy-ws/src/SmithyRsSource/codegen-server-test/model/simple.smithy
     |
  36 |     payload: StringList
     |     ^
     = AWS Protocols do not support applying the httpPayload trait to members that target sets, lists, or maps.
```

Which I didn't know, but it's called out in the protocol e.g. [restJson1](https://awslabs.github.io/smithy/1.0/spec/aws/aws-restjson1-protocol.html#supported-traits). We only offer support for AWS protocols:

```
Projection simple failed: software.amazon.smithy.codegen.core.CodegenException: No matching protocol ? service offers: []. We offer: [aws.protocols#awsJson1_0, aws.protocols#awsJson1_1, aws.protocols#awsQuery, aws.protocols#ec2Query, aws.protocols#restJson1, aws.protocols#restXml]
```

So I'll just leave a comment documenting this fact.

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpProtocolGenerator.kt (hunk 1 above)
Comment (david-perez) [reply codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpProtocolGenerator.kt:574]: You mean the type of the function `RustWriter.(String) -> Unit`? Type inference fails if I omit the signature.

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/generators/http/HttpBindingGenerator.kt (hunk 6 above)
Comment (david-perez) [reply codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/generators/http/HttpBindingGenerator.kt:91]: I moved `ResponseBindingGenerator.kt` to `HttpBindingGenerator.kt`; all the existing comments use the imperative. I'll add the full stops though.

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpProtocolGenerator.kt (hunk 1 above)
Comment (crisidev) [reply codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpProtocolGenerator.kt:574]: Yep, that's what I meant. Thanks for thee clarification.

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/generators/http/HttpBindingGenerator.kt (hunk 3 above)
Comment (david-perez) [reply codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/generators/http/HttpBindingGenerator.kt:62]: All these are sentences split across bullet points. The style is:

```
* item 1,
* item 2,
* ...
* item n-1; and
* item n.
```

where `and` can of course be any other grammatical conjunction.

Or:

```
* item 1,
* item 2,
* ...
* item n-1; and
* item n

if the sentence continues past the bullet points and ends in the next line.
```

I don't think there's an agreed upon standard, but some formal / technical writing does use this style.

I don't have strong opinions on this, I just try to be consistent.

**Review by crisidev — APPROVED** [review]
LGTM

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/generators/http/HttpBindingGenerator.kt (hunk 3 above)
Comment (crisidev) [reply codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/generators/http/HttpBindingGenerator.kt:62]: I have no strong opinions as well. It just feels weird to me, but not a blocker. Same for the rest below.

**Review by rcoh — APPROVED** [review]
httpbinding refactoring lgtm

**Comment by guymguym** [comment]
Hi @david-perez 
Is there any overlaps with my PR #1023 ? 
Should I change anything? Thanks

**Comment by david-perez** [comment]
@guymguym Ah yes, the logic for setting headers that you copied over in your PR should be moved to the class `HttpBindingGenerator.kt` this PR introduces, since it's useful for both clients and servers. That would resolve my comment here https://github.com/awslabs/smithy-rs/pull/1023#discussion_r777684656

I also need to set headers in responses, to resolve #1071 and #1075, so I'll merge this PR and work on moving header-setting to `HttpBindingGenerator.kt` so that you'd only have to rebase your PR. I'll start today.

**Pull request merged**